JAVA-983: [3.0] QueryBuilder cannot handle collections containing function calls

adutra opened the pull request:
I'm opening this PR against 3.0 but it should be backported to 2.1 as well.

beckje01 committed:
JAVA-968: Update metrics to the latest version.

Latest version changed groupId to io.dropwizard.metrics from com.codahale.metrics.

Commit:
Use all resolved InetAddresses for contact points.

This will greatly simplify setups as we can add a single cassandra_seeds
dns record and have that statically added to our configuration.

olim7t committed:
Bump to next snapshot version for 2.2.

olim7t committed:
Merge branch '2.1' into 2.2

Commit:
Protocol V4 enum support (JAVA-783).

olim7t committed:
Expose function and aggregate metadata (JAVA-747).

olim7t committed:
Fix table and UDT parsing on whole schema refresh (JAVA-782).

Commit:
Add k/v payload for 3rd party usage (JAVA-779).

olim7t committed:
Handle function and aggregate schema change events/results (JAVA-775).

snazy committed:
Add support for smallint and tinyint (JAVA-786)

olim7t committed:
Add new client exception for CQL function failure (JAVA-778).

snazy committed:
Support new C* 2.2 CQL date and time types (JAVA-404).

olim7t committed:
Use PK columns in protocol v4 PREPARED response (JAVA-776).

Commit:
Distinguish NULL and UNSET values (JAVA-777).

olim7t committed:
Expose server-side warnings on ExecutionInfo (JAVA-780).

olim7t committed:
Expose new read/write failure exceptions (JAVA-749).

olim7t committed:
Remove deprecated methods.

olim7t committed:
Rename getProtocolVersionEnum to getProtocolVersion.

olim7t committed:
Rename AddressTranslater to AddressTranslator.

olim7t committed:
Merge CloseableLoadBalancingPolicy into LoadBalancingPolicy.

olim7t committed:
Merge CloseableAddressTranslator into AddressTranslator.

olim7t committed:
Add cluster lifecycle awareness to all pluggable components.

- *Policy, AddressTranslator, TimestampGenerator now have init(Cluster)
  and close()
- LatencyTracker and StateListener now have onRegister(Cluster) and
  onUnregister(Cluster)
- LoadBalancingPolicy and ControlConnection do not extend StateListener
  anymore
- updated QueryLogger to use register.

tolbertam committed:
Integration test improvements for JAVA-783.

- Introduce TestUtils.getDesiredProtocolVersion(maxAllowed) for
  retrieving a ProtocolVersion with the maximum returned being the one
  provided.  Update DynamicConnectionPoolTest and some tests in
  PreparedStatementTest to use it.
- Add WriteFailureException to ExceptionsTest.
- Move test in ReadFailureTest to ExceptionsTest.
- Remove UnsetValuesTest as all of these tests are now present in
  PreparedStatementTest.
- Fix LargeDataTest#testWideRows so it writes up the default max allowed
  batch size in 2.2.0-beta1.

Commit:
Fixing wrong check for supplied parameter heartbeatIntervalSeconds.

olim7t committed:
Merge branch '2.1' into 2.2

adutra committed:
Merge pull request #346 from datastax/java783

Protocol V4 enum support (JAVA-783).

adutra committed:
Merge pull request #351 from snazy/786-small-tiny

Add support for smallint and tinyint (JAVA-786).

tolbertam committed:
Add DataType.allPrimitiveTypes(ProtocolVersion).

With the addition of tinyint, smallint, date and time we now have a case
where some primitive types are not supported in all native protocol
versions.  Therefore it seems useful to be able to request primitive
types by version.

This method is currently only exposed as package private for internal
use.  Updated tests to use this method in favor of
DataType.allPrimitiveTypes().

tolbertam committed:
Merge pull request #359 from tolbertam/DataTypeMinProtocolVersion

Add DataType.allPrimitiveTypes(ProtocolVersion).

olim7t committed:
Make client timestamps the default (with atomic generator).

tolbertam committed:
Selectively enable user defined functions based on cassandra version.

olim7t committed:
Merge pull request #348 from datastax/udf

User-defined functions in 2.2

olim7t committed:
Expose ProtocolVersion#toInt (JAVA-700).

olim7t committed:
Merge pull request #360 from datastax/java616

API cleanup for 2.2 (JAVA-616)

olim7t committed:
Additional changes on new DateWithoutTime type (JAVA-404).

olim7t committed:
Add link to platform survey.

olim7t committed:
Merge pull request #365 from datastax/java404

Add date+time support (JAVA-404)

adutra committed:
Merge pull request #350 from datastax/java779

Add k/v payload for 3rd party usage (JAVA-779)

olim7t committed:
Update upgrade guide.

olim7t committed:
Fix reference + list new features in native protocol doc.

olim7t committed:
Update javadoc links in docs.

olim7t committed:
Update query timestamps doc now that client-side is the default.

olim7t committed:
Explain offset queries in paging doc.

tolbertam committed:
Temporarily have CustomPayloadTest require 2.2.1 as maven still runs it.

tolbertam committed:
Split MapperPrimitiveTypesTest into 2 separate tests for compatibility

tolbertam committed:
Add license header.

Commit:
Updating version numbers in README files.

Commit:
Updating docs.yaml.

Commit:
[maven-release-plugin] prepare release 2.2.0-rc1

Commit:
[maven-release-plugin] prepare for next development iteration

Commit:
Updating docs.yaml.

Commit:
JAVA-810: Rename DateWithoutTime to LocalDate.

Commit:
JAVA-876: Support new system tables in C* 3.0.0-alpha1.

olim7t committed:
Merge pull request #373 from datastax/java810

JAVA-810: Rename DateWithoutTime to LocalDate.

Commit:
Allow user to register custom type codecs (JAVA-721, JAVA-722).

Commit:
JAVA-818: TypeCodec.getDataTypeFor() does not handle LocalDate instances.

Commit:
DateCodec does not format values correctly (JAVA-816) - TimeCodec does not format values correctly (JAVA-817).

Commit:
Bump Cassandra version to 2.2.0-rc1.

Commit:
Fix failing test.

olim7t committed:
Add 2.1.7 to generated docs.

olim7t committed:
Merge branch '2.1' into 2.2

Commit:
Prevent clients from using protocol V4 when connecting to a Scassandra server.

With V4, the server fails to properly detect the protocol version sent by the client and consequently fails to interpret the very first message received.

Commit:
Amend TupleTest to fix a failing test involving nested tuples.

With C* 2.2 + (and JDK 8 consequently), a type containing 128 nested tuples causes the server to crash with a StackOverflowError. Lowering the number of nested tuples to 64 solves the problem.

Commit:
Refactor SchemaTest to make tests compatible with C* 2.2 + and JDK 8.

Some tests in the old SchemaTest class were failing due to the use of JDK 8 instead of JDK 6. For example, the order in which entries in a HashMap are printed is not the same, and some assertions were relying on a specific order.
Moreover, these tests were fragile they were relying on the equality between the CREATE TABLE statement and the string produced by TableMetadata.exportAsString().

Commit:
Bump Cassandra version from 2.2.0-rc1 to 2.2.0-rc2.

Commit:
Fix wrong data center names for multi-DC tests.

tolbertam committed:
Don't set enable_user_defined_functions unless 2.2+.

olim7t committed:
Make ResultSet#fetchMoreResult return a ListenableFuture<ResultSet> (JAVA-836)

olim7t committed:
Merge pull request #386 from datastax/java816

JAVA-816, JAVA-817, JAVA-818

Commit:
Enable custom payload tests.

olim7t committed:
Merge branch '2.1' into 2.2

olim7t committed:
Disable frozen checks in mapper (JAVA-843)

olim7t committed:
Merge pull request #411 from datastax/java836

Make ResultSet#fetchMoreResult return a ListenableFuture<ResultSet> (JAVA-836)

olim7t committed:
Merge pull request #412 from datastax/java843

Disable frozen checks in mapper (JAVA-843)

olim7t committed:
Update version in docs.

adutra committed:
Merge pull request #387 from datastax/java721-2.2

Allow user to register custom type codecs (JAVA-721, JAVA-722).

Commit:
[maven-release-plugin] prepare release 2.2.0-rc2

Commit:
[maven-release-plugin] prepare for next development iteration

Commit:
Fix broken links in documentation.

Commit:
Include 2.2.2-rc2 version in docs.yaml.

Commit:
JAVA-847: CodecRegistry is not propagated to nested UDTs.

Commit:
JAVA-848: Ability to store a default, shareable CodecRegistry instance.

tolbertam committed:
1.2 compatibility in TypeCodecArraysIntegrationTest.

Commit:
Update versions to 3.0.0-alpha1-SNAPSHOT.

Commit:
Update Cassandra version to 3.0.0-alpha1

Commit:
Update main README file.

Commit:
Update comparison version for clirr plugin to 3.0.0.

adutra committed:
Merge pull request #429 from datastax/java876

JAVA-876: Support new system tables in C* 3.0.0-alpha1.

Commit:
[maven-release-plugin] prepare release 3.0.0-alpha1

Commit:
[maven-release-plugin] prepare for next development iteration

Commit:
Include 3.0.0-alpha1 version in docs.yaml.

Commit:
Move secondary index metadata out of column definitions (JAVA-875, JAVA-882).

Commit:
Avoid multi query pagination with C* 3.0 and fetchSize = 1.

Commit:
JAVA-880: TupleCodec and UDTCodec should treat empty ByteBuffers as empty values, not null values.

olim7t committed:
Merge pull request #414 from datastax/java847

JAVA-847: CodecRegistry is not propagated to nested UDTs.

olim7t committed:
Merge pull request #415 from datastax/java848

JAVA-848: Ability to store a default, shareable CodecRegistry instance.

olim7t committed:
Merge pull request #432 from datastax/java880

JAVA-880: TupleCodec and UDTCodec should treat empty ByteBuffers as empty values, not null values.

olim7t committed:
Provide "disconnected" constructors for QueryBuilder and TupleType.

olim7t committed:
Improve javadoc of SimpleStatement.

tolbertam committed:
Add/update tests for 'detached' TupleType/QueryBuilder instances.

bdeggleston committed:
Add test for indexes on different tables (JAVA-882).

olim7t committed:
Temporarily disable test that fails with Cassandra 3.

olim7t committed:
Merge branch '2.2' into 3.0

olim7t committed:
Fix changelog.

mike-tr-adamson committed:
JAVA-885: Pass server authenticator class through to new authenticator.

olim7t committed:
Update version in documentation.

olim7t committed:
[maven-release-plugin] prepare release 2.2.0-rc3

olim7t committed:
[maven-release-plugin] prepare for next development iteration

olim7t committed:
Add 2.2.0-rc3 to docs.

olim7t committed:
Merge pull request #430 from datastax/java875

JAVA-875: Move secondary index metadata out of column definitions.

olim7t committed:
Update version in docs.

olim7t committed:
[maven-release-plugin] prepare release 3.0.0-alpha2

olim7t committed:
[maven-release-plugin] prepare for next development iteration

Commit:
JAVA-912: Integrate changes to 'columns' table.

olim7t committed:
Add 2.0.11 to the doc.

olim7t committed:
Switch doc to Datastax theme.

Commit:
JAVA-913: Redesign IndexMetadata API.

This commit contains substantial contributions by Andrew Tolbert (@tolbertam).

Commit:
JAVA-825: Parse materialized views metadata.

adutra committed:
Merge pull request #453 from datastax/java825

JAVA-825: Parse materialized views metadata.

Commit:
JAVA-936: Adapt schema metadata parsing logic to new storage format of CQL types in C* 3.0.

This commit also migrates the 'initcond' column type in table
'system_schema.aggregates' from blob to varchar (see CASSANDRA-10650).

Commit:
JAVA-918: Allow MV's select statements to be more complex.

olim7t committed:
Merge branch '2.1' into 2.2

tolbertam committed:
Make ccmBridge in PerClassSingleNodeCluster private.

Changed ccmBridge to private and added 'stop' and 'start' methods on
PerClassSingleNodeCluster for stopping/starting the node while
preserving jvm arguments.

Commit:
JAVA-912: Support for new 'clustering_order' column.

Commit:
JAVA-911: Process materialized views events.

Commit:
JAVA-919: Move crc_check_chance out of compressions options.

adutra committed:
Merge pull request #452 from datastax/java913

JAVA-913: Redesign IndexMetadata API.

adutra committed:
Merge pull request #459 from datastax/java911

JAVA-911: Process materialized views events.

adutra committed:
Merge pull request #458 from datastax/java912b

JAVA-912: Support for new 'clustering_order' column.

tolbertam committed:
Add should_parse_new_compression_options() test for CASSANDRA-9424.

olim7t committed:
Improve equals for Metadata classes.

olim7t committed:
Merge branch '2.2' into 3.0

Commit:
Ninja-fix bug when refreshing materialized views during a partial TABLE refresh.

When refreshing only a table or only a materialized view, the CQL query generated would be invalid with a WHERE clause containing "view_name = 'null'".

tolbertam committed:
Verify on empty list signature.

tolbertam committed:
Merge pull request #460 from datastax/java919

JAVA-919: Move crc_check_chance out of compressions options.

tolbertam committed:
Accessor tests for materialized views.

iamaleksey committed:
Merge pull request #456 from datastax/java918

JAVA-918: Allow MV's select statements to be more complex.

Commit:
Bumping versions to 3.0.0-alpha3.

Commit:
[maven-release-plugin] prepare release 3.0.0-alpha3

Commit:
[maven-release-plugin] prepare for next development iteration

Commit:
Make table metadata options class visible (JAVA-946).

Commit:
JAVA-922: Make TypeCodec return mutable collections.

This commit relaxes the postconditions on immutability stated in the javadocs for some methods in these classes.
This commit also updates the javadocs of GettableByIndexData and GettableByNameData accordingly.
It also prunes Row class and removes all duplicated method declarations.
The Upgrade guide is updated with behavioral changes introduced by this commit.

Commit:
Use Collections.nCopies() to initialize lists.

olim7t committed:
Update Twitter accounts.

olim7t committed:
Update Twitter accounts.

Commit:
JAVA-877: Don't wait NEW_NODE_DELAY_SECONDS if protocol version = V4.

olim7t committed:
Add 2.1.8 to docs.

tolbertam committed:
JAVA-939 Add crcCheckChance to TableOptionsMetadata#equals/hashCode.

olim7t committed:
Fix Configuration builder to allow disabled metrics (JAVA-900).

olim7t committed:
Merge AsyncInitSession into Session.

Commit:
JAVA-932: Limit visibility of codec internals.

- do not expose default codec classes
- remove "utility" codecs from public API (Enum*Codec, MappingCodec,
  StringParsingCodec).
- remove setObject method from SettableBy*Data.

olim7t committed:
Add BoundStatement#unset (JAVA-930).

Commit:
Allow typed getters/setters to target any CQL type (JAVA-940).

olim7t committed:
Merge pull request #461 from datastax/schema-debouncer-fix

Verify on empty list signature in should_debounce_and_coalesce_multiple_alter_events_on_same_table_into_refresh_table

olim7t committed:
Merge pull request #457 from datastax/custom_payload_test_fix

Fix CustomPayloadTest failing when run in suite.

Commit:
JAVA-942: Fix implementation of UserType.hashCode()

olim7t committed:
Warn if a custom codec collides with an existing one (JAVA-934).

olim7t committed:
Prepare API for async query trace (JAVA-902).

tolbertam committed:
JAVA-938 Parse extensions into TableOrView Options.

olim7t committed:
Merge branch '2.2' into 3.0

olim7t committed:
Change default consistency level to LOCAL_QUORUM (JAVA-926).

olim7t committed:
Merge branch '2.1' into 3.0

olim7t committed:
Merge branch 'java926' into 3.0

olim7t committed:
Merge pull request #483 from datastax/java942

JAVA-942: Fix implementation of UserType.hashCode()

olim7t committed:
Merge pull request #482 from datastax/java877

JAVA-877: Don't wait NEW_NODE_DELAY_SECONDS if protocol version = V4.

olim7t committed:
Merge pull request #478 from datastax/java938

JAVA-938 Parse extensions into TableOrView Options.

tolbertam committed:
Add 2.2 and 3.0 to build.yaml

tolbertam committed:
Add tests for Metrics enablement/disablement.

tolbertam committed:
Reorder CacheBuilder construction for better JDK6 support.

tolbertam committed:
Test fixes for Codec changes.

tolbertam committed:
Fix DataType check in TimeZonePreservingDateTimeCodec.

olim7t committed:
Merge pull request #481 from datastax/java900

Fix Configuration builder to allow disabled metrics (JAVA-900).

tolbertam committed:
Add GettableData/SettableData tests for JAVA-940.

olim7t committed:
Fix Cluster.connect with a case-sensitive keyspace (JAVA-950).

Go back to the previous strategy of first initializing a session with no
keyspace and then issuing a USE statement. This guarantees that
SessionManager.poolsState.keyspace always contains a keyspace in the
"internal" format (unquoted and in the correct case), and keeps the
implementation simple.
The downside is that if the keyspace is wrong, we'll open all
connections before realizing the error, but that's an edge case.

olim7t committed:
Merge pull request #480 from datastax/java902

Prepare API for async query trace (JAVA-902).

olim7t committed:
Merge pull request #479 from datastax/java930

Add BoundStatement#unset (JAVA-930).

olim7t committed:
Merge pull request #463 from datastax/java825b

Make TableOrView public to allow TableOrView.Option to be referenced outside package.

olim7t committed:
Merge pull request #470 from datastax/java939

JAVA-939 Add crcCheckChance to Options#equals/hashCode.

olim7t committed:
Merge pull request #485 from datastax/java931

TypeCodec improvements (JAVA-931)

olim7t committed:
Update codec-related Javadocs.

olim7t committed:
Merge pull request #486 from datastax/java950_3.0

Fix Cluster.connect with a case-sensitive keyspace (JAVA-950).

olim7t committed:
Fix MaterializedViewMetadata when base table name is case sensitive (JAVA-953).

tolbertam committed:
Update OSGi example Guava versions. Refine bundle version detection.

olim7t committed:
Merge pull request #487 from datastax/java953

Fix MaterializedViewMetadata when base table name is case sensitive (…

olim7t committed:
Update version in docs.

Commit:
JAVA-957: Improve Custom Codecs documentation on Type Inheritance

olim7t committed:
Merge pull request #491 from datastax/java957

JAVA-957: Improve Custom Codecs documentation on Type Inheritance

olim7t committed:
[maven-release-plugin] prepare release 3.0.0-alpha4

olim7t committed:
[maven-release-plugin] prepare for next development iteration

Commit:
Add version 3.0.0-alpha4 to docs.yaml.

Commit:
JAVA-971: Make type codecs invariant.

This commit makes TypeCodec.accepts(TypeToken) invariant with respect to the Java type passed as argument.
TypeCodec.accepts(Object) remains covariant with respect to the runtime type of the passed object.

tolbertam committed:
JAVA-965: Improve error handling of non-type 1 UUID used for timeuuid.

This commit also applies the same behavior to AsciiCodec,
when given a non-ASCII string.

tolbertam committed:
JAVA-961: Raise an excepiton when an older version of guava is found.

Commit:
Merge branch 'beckje01-metricsUpgrade' into 3.0

Commit:
JAVA-958: Make TableOrView.Order visible.

This commit promotes the enum Order to a public top level class and renames it to ClusteringOrder.

Commit:
JAVA-972: TypeCodec.parse() implementations should be case insensitive when checking for keyword NULL.

adutra committed:
Merge pull request #496 from datastax/java958

JAVA-958: Make TableOrView.Order visible.

adutra committed:
Merge pull request #494 from datastax/osgiExampleUpdate-3.0

Update OSGi example guava versions. Refine bundle version detection.

olim7t committed:
JAVA-841, JAVA-948: Refactor SSLOptions API.

olim7t committed:
JAVA-917: Document SSL configuration.

adutra committed:
Merge pull request #492 from datastax/java965

JAVA-965 Improve error handling of non-type 1 UUID used for timeuuid.

adutra committed:
Merge pull request #503 from datastax/java885

JAVA-885: Pass server authenticator class through to new authenticator.

tolbertam committed:
Merge pull request #493 from datastax/java961

JAVA-961: Raise an exception when an older version of guava (< 16.0.1) is found.

adutra committed:
Merge pull request #511 from datastax/java972

JAVA-972: TypeCodec.parse() implementations should be case insensitive when checking for keyword NULL.

Commit:
JAVA-983: QueryBuilder cannot handle collections containing function calls.

This commit also incidentally formats INSERT statements with a whitestpace between table name and column list,
so that they do not look like a function call.

olim7t committed:
Update docs.yaml for 2.0.12 and 2.1.9.

tolbertam committed:
JAVA-986: Update documentation links to reference 3.0.

Commit:
Merge branch '2.1' into 3.0

Conflicts:
	README.md
	changelog/README.md
	driver-core/pom.xml
	driver-core/src/main/java/com/datastax/driver/core/Cluster.java
	driver-core/src/main/java/com/datastax/driver/core/ClusteringOrder.java
	driver-core/src/main/java/com/datastax/driver/core/Connection.java
	driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java
	driver-core/src/main/java/com/datastax/driver/core/SetKeyspaceException.java
	driver-core/src/main/java/com/datastax/driver/core/UserType.java
	driver-core/src/main/java/com/datastax/driver/core/policies/WhiteListPolicy.java
	driver-core/src/test/java/com/datastax/driver/core/CCMBridge.java
	driver-core/src/test/java/com/datastax/driver/core/HostAssert.java
	driver-core/src/test/java/com/datastax/driver/core/ReconnectionTest.java
	driver-core/src/test/java/com/datastax/driver/core/SCassandraCluster.java
	driver-core/src/test/java/com/datastax/driver/core/SimpleStatementTest.java
	driver-core/src/test/java/com/datastax/driver/core/StateListenerBase.java
	driver-dist/pom.xml
	driver-dse/pom.xml
	driver-examples/osgi/pom.xml
	driver-examples/pom.xml
	driver-examples/stress/pom.xml
	driver-mapping/pom.xml
	driver-mapping/src/test/java/com/datastax/driver/mapping/MapperTest.java
	features/compression/README.md
	features/object_mapper/README.md
	features/object_mapper/creating/README.md
	features/shaded_jar/README.md
	pom.xml

adutra commented on driver-core/src/main/java/com/datastax/driver/core/querybuilder/Insert.java:
```diff
@@ -56,7 +56,7 @@ StringBuilder buildQueryString(List<Object> variables) {
             Utils.appendName(keyspace, builder).append('.');
         Utils.appendName(table, builder);
-        builder.append('(');
+        builder.append(" (");
```
Not required, but imho the resulting string looks better with this extra whitespace between the table name and the columns list.

olim7t replied:
Agreed, that's also how insert statements are formatted in the CQL reference doc.
IMO this should stay as a separate commit when we eventually merge the PR.

tolbertam commented:
Shouldn't there also be some kind of handling for Tuples and UDTs?

adutra commented:
I don't think so; TupleValue and UDTValue instances cannot contain special values such as function calls; their API is that of SettableData and as such setting a UDT or tuple field to a function call is not possible.

tolbertam commented:
Ah I see, thanks for clearing that up.  For both UDTValue and TupleValue the types must be expressed beforehand, so there would be no way to say 'fcall("minTimeUuuid", "2013-02-03 10:00+0000")' in a UDTValue or TupleValue containing a timeuuid.

tolbertam commented:
Looks good to me :+1:

Commit:
Update changelog format.

This commit introduces the following new format :
- [type] JAVA-XXX: description.

olim7t commented:
:+1:
I'm just wondering if we shouldn't fix this first on 2.1 and then merge downstreams, backporting commits is always a bit clunky.
The same applies to #523.

Commit:
Improve javadocs of Cluster.register() methods.

adutra committed:
Merge pull request #504 from datastax/java971

JAVA-971: Make type codecs invariant.

adutra committed:
Merge pull request #518 from datastax/java986

JAVA-986: Update documentation links to reference 3.0.

adutra committed:
Merge pull request #508 from datastax/java841

Refactor SSLOptions API (JAVA-841, JAVA-948).

olim7t committed:
Merge branch '2.1' into 3.0

adutra commented:
Squashed and rebased on top of current 3.0.

adutra committed:
Merge pull request #467 from datastax/java936

JAVA-936: Adapt schema metadata parsing logic to new storage format of CQL types in C* 3.0

tolbertam committed:
Update tests to only use netty openssl on JDK 1.7+ and add note to docs.

tolbertam committed:
Merge pull request #527 from datastax/nettySslJdk7Requirement

Update tests to only use netty openssl on JDK 1.7+ and add note to docs.

Commit:
Merge branch '2.1' (early part) into 3.0

Commit:
Merge branch '2.1' into 3.0
